### PR TITLE
hydra-core 1.3.2 :snowflake:

### PR DESCRIPTION
**Pull request by boldorider4**
hydra-core 1.3.2 :snowflake:

**Destination channel:** Snowflake

### Links

- [PKG-5660]
- dev_url:        https://github.com/facebookresearch/hydra/tree/v1.3.2
- conda_forge:    https://github.com/conda-forge/hydra-core-feedstock
- pypi:           https://pypi.org/project/hydra-core/1.3.2
- pypi inspector: https://inspector.pypi.io/project/hydra-core/1.3.2
- Depends on:
     - https://github.com/AnacondaRecipes/antlr-feedstock/pull/1
- Dependency for:
     - fairsq 0.12.2

### Explanation of changes:

- we are pinning an exact version of `antlr`, as the minor version introduces API breakage

[PKG-5660]: https://anaconda.atlassian.net/browse/PKG-5660?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: recipe/0000-unvendor_antlr.patch [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
++        root_dir = os.environ['PREFIX']
+         project_root = abspath(dirname(basename(__file__)))
+         for grammar in [
+             "hydra/grammar/OverrideLexer.g4",
+@@ -185,7 +185,7 @@
+             command = [
+                 "java",
+                 "-jar",
+-                join(root_dir, "bin/antlr-4.9.3-complete.jar"),
++                join(root_dir, "lib/antlr4-4.9.3-complete.jar"),
```
Comment (lorepirri) [inline recipe/0000-unvendor_antlr.patch]: could you try with:

```
os.environ['PKG_VERSION']
```

?

something like:

```suggestion
+                join(root_dir, f"lib/antlr4-{os.environ['PKG_VERSION']}-complete.jar"),
```

File: abs.yaml [hunk 2]
```diff
@@ -0,0 +1,4 @@
+channels:
+- https://staging.continuum.io/prefect/fs/antlr-feedstock/pr1/a7543c5
+- sfe1ed40
```
Comment (lorepirri) [inline abs.yaml]:
```suggestion
upload_channels:
  - sfe1ed40

channels:
  - https://staging.continuum.io/prefect/fs/antlr-feedstock/pr1/a7543c5
```

File: recipe/meta.yaml [hunk 3]
```diff
@@ -1,58 +1,70 @@
...
   version: {{ version }}
 
 source:
-  url: https://github.com/facebookresearch/hydra/archive/v{{ version }}.tar.gz
-  sha256: 9ba1efe09893853e883498fcbeec17f2a77e1ba551587d6d0604077e02de1602
+  - url: https://pypi.io/packages/source/{{ name[0] }}/{{ name }}/{{ name }}-{{ version }}.tar.gz
+    sha256: 8a878ed67216997c3e9d88a8e72e7b4767e81af37afb4ea3334b269a4390a824
+    patches:
+    # do not forget to update the version literal in the antlr-*-complete.jar string inside the
+    # patch to correctly unvendor antlr and use the antlr package from conda
```
Comment (lorepirri) [inline recipe/meta.yaml]: let's try to use `os.environ['PKG_VERSION']` instead

```suggestion
```

File: recipe/meta.yaml [hunk 4]
```diff
@@ -1,58 +1,70 @@
...
+      - 0000-unvendor_antlr.patch
+    folder: pypi-upstream
```
Comment (lorepirri) [inline recipe/meta.yaml]: why a different folder?

File: recipe/meta.yaml [hunk 5]
```diff
@@ -1,58 +1,70 @@
...
     - pip
+    - openjdk
   run:
-    - python >=3.6
+    - python
     - omegaconf >=2.2,<2.4
-    - antlr-python-runtime ==4.9.*
+    # minor versions are introducing breaking changes (another version was tried, e.g. 4.13.* and implementation
+    # turned out to be incompatible)
+    - antlr4-python3-runtime >=4.9,<4.10
```
Comment (lorepirri) [inline recipe/meta.yaml]:
```suggestion
    - antlr4-python3-runtime >=4.9,<4.10.0a0
```

or you could leave:

```suggestion
    - antlr4-python3-runtime ==4.9.*
```
https://github.com/facebookresearch/hydra/blob/7ef3acab78b46bf82370e7931f748aef5fd80e5f/requirements/requirements.txt#L2

File: recipe/meta.yaml [hunk 6]
```diff
@@ -1,58 +1,70 @@
...
+    - {{ PYTHON }} -m pip install $SRC_DIR/pypi-upstream --no-deps --no-build-isolation -vv
 
 requirements:
   build:
-    - antlr ==4.9.*
+    - patch  # [not win]
+    - m2-patch  # [win]
   host:
-    - python >=3.6
+    - antlr 4.9.3
```
Comment (lorepirri) [inline recipe/meta.yaml]: why was this moved to host? will it pull in `openjdk` or must we have it here? (line 33)

```suggestion
    - antlr ==4.9.*
```

or

```suggestion
    - antlr >4.9.0,<4.10.0a0
```

File: recipe/0000-unvendor_antlr.patch (hunk 1 above)
Comment (psteyer) [reply recipe/0000-unvendor_antlr.patch]: This seems to be breaking the build:

```
raise CalledProcessError(retcode, cmd)
  subprocess.CalledProcessError: Command '['java', '-jar', '/opt/conda/conda-bld/hydra-core_1725954885436/_h_env_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_placehold_pl/lib/antlr4-1.3.2-complete.jar', '-Dlanguage=Python3', '-o', '/opt/conda/conda-bld/hydra-core_1725954885436/work/pypi-upstream/hydra/grammar/gen/', '-Xexact-output-dir', '-visitor', '/opt/conda/conda-bld/hydra-core_1725954885436/work/pypi-upstream/hydra/grammar/OverrideLexer.g4']' returned non-zero exit status 1.
  error: subprocess-exited-with-error
```


```
ERROR: Failed building wheel for hydra-core
Failed to build hydra-core
ERROR: ERROR: Failed to build installable wheels for some pyproject.toml based projects (hydra-core)
```

Comment (psteyer) [reply recipe/0000-unvendor_antlr.patch]: I believe we were running into a similar issue which is why we hardcore the version in.

Comment (lorepirri) [reply recipe/0000-unvendor_antlr.patch]: `os.environ['PKG_VERSION']` did not work because it's the actual package version (`1.3.2`) and not the version of `antlr` that we want to use. I pushed a change that uses an ad-hoc environment variable `ANTLR_VERSION` set in the recipe. The version of `antlr` and `antlr4-python3-runtime` should be in sync anyway.

File: recipe/meta.yaml (hunk 4 above)
Comment (lorepirri) [reply recipe/meta.yaml]: could you please use the root folder for this? (remove the `folder` tag, and update the `license_file` path)

Comment (boldorider4) [reply recipe/meta.yaml]: this was due to an old suggestion from @JeanChristopheMorinPerso, that whenever we have both a PyPi and a GitHub tarball, because we give priority to the PyPi for the actual package build and only fetch the GitHub tarball to have some test assets that are not present in the PyPi archive, we put these in separate directories to prevent the directory where the GitHub archive gets extracted to overwrite a directory in the PyPi archive, which would end up in the same path.
Correct me if you stand by this @JeanChristopheMorinPerso. I may have misunderstood.

I guess that if we check the content of the PyPi archive and give the `test-assets` directory a proper name that doesn't conflict, we shouldn't have to worry about this, but at the same time this way we will always be sure of not running into conflicting names.

File: recipe/0000-unvendor_antlr.patch (hunk 1 above)
Comment (boldorider4) [reply recipe/0000-unvendor_antlr.patch]: Awesome! Thanks @lorepirri!

File: recipe/meta.yaml (hunk 3 above)
Comment (boldorider4) [reply recipe/meta.yaml]: I will remove the comment.

File: recipe/meta.yaml (hunk 4 above)
Comment (lorepirri) [reply recipe/meta.yaml]: It would be enough to have the GH archive for the tests + data files in a folder such as gh_src and the pypi archive in the root. You would avoid using `%SRC_DIR%` for win and `$SRC_DIR` for the rest.

**Comment by lorepirri** [comment]
@boldorider4 @psteyer if you:

- revert to https://github.com/AnacondaRecipes/hydra-core-feedstock/pull/2/commits/72e53c233a87120a211ba7a0ec3d135200e0290d 
- and avoid using the `folder: pypi-upstream`

as suggested in https://github.com/AnacondaRecipes/hydra-core-feedstock/pull/2#discussion_r1754516629 and https://github.com/AnacondaRecipes/hydra-core-feedstock/pull/2#discussion_r1753506476

everything will be working fine, also for `win`.

**Comment by lorepirri** [comment]
In the meantime, @psteyer you could be working on https://github.com/AnacondaRecipes/fairseq-feedstock/pull/1 using the staging channel to this package `hydra-core` at commit id 72e53c2 because it built successfully except for `win` because of the `folder: pypi-upstream` usage.

**Review by psteyer — APPROVED** [review]
LGTM

**Review by lorepirri — APPROVED** [review]
self-approving

File: recipe/meta.yaml [hunk 7]
```diff
@@ -1,58 +1,91 @@
...
 
-    # importlib_resources is only needed for Python <3.9, but can't
-    # specify that with `noarch`
-    - importlib_resources
+# list tests to be skipped
+{% set tests_to_skip = "" %}
+# skip all tests on win that use hard coded paths because of slash backslash
+{% set tests_to_skip = tests_to_skip + " test_hydra_output_dir" %}                       # [win]
+{% set tests_to_skip = tests_to_skip + " or test_specializing_config_example" %}         # [win]
+{% set tests_to_skip = tests_to_skip + " or test_1" %}
```
Comment (ifitchet) [inline recipe/meta.yaml]: Missing a selector?

Comment (lorepirri) [reply recipe/meta.yaml]:
```suggestion
{% set tests_to_skip = tests_to_skip + " or test_1" %}  # [win]
```

yes! Thanks
